Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,165 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Connects to a SOCKS server to establish a proxied connection
+/// to a host. This handler should be inserted at the beginning of a
+/// channel's pipeline. Note that SOCKS only supports fully-qualified
+/// domain names and IPv4 or IPv6 sockets, and not UNIX sockets.
+public final class SOCKSClientHandler: ChannelDuplexHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+    
+    private let targetAddress: SOCKSAddress
+    
+    private var state: ClientStateMachine
+    private var buffered: ByteBuffer?
+    
+    private var bufferedWrites: CircularBuffer<(NIOAny, EventLoopPromise<Void>?)> = .init()
+    
+    /// Creates a new `SOCKSClientHandler` that connects to a server
+    /// and instructs the server to connect to `targetAddress`.
+    /// - parameter targetAddress: The desired end point - note that only IPv4, IPv6, and FQDNs are supported.
+    public init(targetAddress: SOCKSAddress) {
+        
+        switch targetAddress {
+        case .address(.unixDomainSocket):
+            preconditionFailure("UNIX domain sockets are not supported.")
+        case .domain, .address(.v4), .address(.v6):
+            break
+        }
+        
+        self.state = ClientStateMachine()
+        self.targetAddress = targetAddress
+    }
+    
+    public func channelActive(context: ChannelHandlerContext) {
+        self.beginHandshake(context: context)
+    }
+    
+    public func handlerAdded(context: ChannelHandlerContext) {
+        self.beginHandshake(context: context)
+    }
+    
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        
+        // if we've established the connection then forward on the data
+        if self.state.proxyEstablished {
+            context.fireChannelRead(data)
+            return
+        }
+        
+        var inboundBuffer = self.unwrapInboundIn(data)
+        
+        if self.buffered == nil {
+            self.buffered = context.channel.allocator.buffer(capacity: inboundBuffer.readableBytes)
```

Comment:
Minor optimisation: we can just store the buffer directly and save ourselves the copy.